File: Embassat/Model/CADEMArtist.m
//
//  CADEMArtist.m
//  Embassat
//
//  Created by Joan Romano on 08/03/14.
//  Copyright (c) 2014 Crows And Dogs. All rights reserved.
//

#import "CADEMArtist.h"

@interface CADEMArtist ()

@property (nonatomic, strong) NSDateFormatter *dateFormatter;
@property (nonatomic, strong) NSDate *initialDate;
@property (nonatomic, strong) NSDate *finalDate;

@end

@implementation CADEMArtist

+ (NSDictionary *)JSONKeyPathsByPropertyKey
{
    return @{
             @"name" : @"title",
             @"longDescription" : @"content",
             @"stage" : @"custom_fields.escenari",
             @"date" : @"custom_fields.data",
             @"artistURL" : @"url",
             @"imageURL" : @"thumbnail_images.medium.url",
             @"initialHour" : @"custom_fields.hora_inici",
             @"initialMinute" : @"custom_fields.minut_inici",
             @"finalHour" : @"custom_fields.hora_final",
             @"finalMinute" : @"custom_fields.minut_final"
             };
}

+ (NSValueTransformer *)artistURLJSONTransformer
{
    return [NSValueTransformer valueTransformerForName:MTLURLValueTransformerName];
}

+ (NSValueTransformer *)imageURLJSONTransformer
{
    return [NSValueTransformer valueTransformerForName:MTLURLValueTransformerName];
}

+ (NSValueTransformer *)initialHourJSONTransformer
{
    return [MTLValueTransformer reversibleTransformerWithBlock:^(NSArray *array) {
        return [array firstObject];
    }];
}

+ (NSValueTransformer *)finalHourJSONTransformer
{
    return [MTLValueTransformer reversibleTransformerWithBlock:^(NSArray *array) {
        return [array firstObject];
    }];
}

+ (NSValueTransformer *)stageJSONTransformer
{
    return [MTLValueTransformer reversibleTransformerWithBlock:^(NSArray *array) {
        return [array firstObject];
    }];
}

+ (NSValueTransformer *)initialMinuteJSONTransformer
{
    return [MTLValueTransformer reversibleTransformerWithBlock:^(NSArray *array) {
        return [array firstObject];
    }];
}

+ (NSValueTransformer *)finalMinuteJSONTransformer
{
    return [MTLValueTransformer reversibleTransformerWithBlock:^(NSArray *array) {
        return [array firstObject];
    }];
}

+ (NSValueTransformer *)dateJSONTransformer
{
    return [MTLValueTransformer reversibleTransformerWithBlock:^(NSArray *array) {
        return [array firstObject];
    }];
}

#pragma mark - Lazy

- (NSDate *)initialDate
{
    return [self realDateWithHour:self.initialHour minute:self.initialMinute];
}

- (NSDate *)finalDate
{
    return [self realDateWithHour:self.finalHour minute:self.finalMinute];
}

- (NSDateFormatter *)dateFormatter
{
    if (!_dateFormatter)
    {
        _dateFormatter = [[NSDateFormatter alloc] init];
        _dateFormatter.dateFormat = @"dd/MM/yyyy HH:mm";
    }
    
    return _dateFormatter;
}

#pragma mark - Private

- (NSDate *)realDateWithHour:(NSString *)hour minute:(NSString *)minute
{
    NSDate *resultDate = [self.dateFormatter dateFromString:[NSString stringWithFormat:@"%@ %@:%@", self.date, hour, minute]];
    
    if ([[hour substringWithRange:NSMakeRange(0, 1)] isEqualToString:@"0"])
    {
        NSDateComponents *compontents = [[NSDateComponents alloc] init];
        compontents.day = 1;
        resultDate = [[NSCalendar currentCalendar] dateByAddingComponents:compontents toDate:resultDate options:0];
    }
    
    return resultDate;
}

@end

File: Embassat/Model/CADEMArtist.swift
//
//  CADEMArtist.swift
//  Embassa't
//
//  Created by Joan Romano on 13/05/15.
//  Copyright (c) 2015 Crows And Dogs. All rights reserved.
//

import Foundation

open class CADEMArtist : NSObject, NSCoding {
    let artistId: Int
    let name: String
    let longDescription: String
    let stage: String
    let youtubeId: String
    let artistURL: URL
    let imageURL: URL
    let startDate: Date
    let endDate: Date
    let scheduleDate: Date
    let scheduleDayString: String
    
    var favorite: Bool
    
    init(artistId: Int, name: String, longDescription: String, artistURL: URL, imageURL: URL, startDate: Date, endDate: Date, stage: String, youtubeId: String, favorite: Bool = false) {
        self.artistId = artistId
        self.name = name
        self.longDescription = longDescription
        self.artistURL = artistURL
        self.imageURL = imageURL
        self.startDate = startDate
        self.endDate = endDate
        self.stage = stage
        self.youtubeId = youtubeId
        self.favorite = favorite    
        self.scheduleDate = startDate.hour < 10 ? startDate.dateBySubstracting(days: 1) : startDate
        self.scheduleDayString = scheduleDate.day == 9 ? "Dijous" : scheduleDate.day == 10 ? "Divendres" : scheduleDate.day == 11 ? "Dissabte" : "Diumenge"
    }
    
    required convenience public init(coder decoder: NSCoder) {
        self.init(artistId: decoder.decodeInteger(forKey: "artistId"),
                  name: decoder.decodeObject(forKey: "name") as! String,
                  longDescription: decoder.decodeObject(forKey: "longDescription") as! String,
                  artistURL: decoder.decodeObject(forKey: "artistURL") as! URL,
                  imageURL: decoder.decodeObject(forKey: "imageURL") as! URL,
                  startDate: decoder.decodeObject(forKey: "startDate") as! Date,
                  endDate: decoder.decodeObject(forKey: "endDate") as! Date,
                  stage: decoder.decodeObject(forKey: "stage") as! String,
                  youtubeId: decoder.decodeObject(forKey: "youtubeId") as! String,
                  favorite: decoder.decodeBool(forKey: "favorite"))
    }
    
    open func encode(with coder: NSCoder) {
        coder.encode(artistId, forKey: "artistId")
        coder.encode(name, forKey: "name")
        coder.encode(longDescription, forKey: "longDescription")
        coder.encode(stage, forKey: "stage")
        coder.encode(youtubeId, forKey: "youtubeId")
        coder.encode(artistURL, forKey: "artistURL")
        coder.encode(imageURL, forKey: "imageURL")
        coder.encode(startDate, forKey: "startDate")
        coder.encode(endDate, forKey: "endDate")
        coder.encode(favorite, forKey: "favorite")
    }
    
    override open var description: String {
        return "Artist: \(name), id \(artistId), description\(longDescription), favorited? \(favorite)"
    }
}
